torch/distributed/fsdp/fully_sharded_data_parallel.py

```diff
@@ -1167,60 +856,335 @@ def _get_ignored_params(
+            )
+            warnings.warn(
+                "Both mixed precision and an `auto_wrap_policy` were specified "
+                "for FSDP, where the wrapped module has batch norm submodules. "
+                "The batch norm submodules will be wrapped as separate FSDP "
+                "instances with mixed precision disabled since some batch norm "
+                "kernels do not support low precision."
+            )
+            auto_wrap_kwargs["auto_wrap_policy"] = auto_wrap_policy
+        _recursive_wrap(**auto_wrap_kwargs, **fsdp_kwargs)
+
+    def _check_single_device_module(
+        self,
+        module: nn.Module,
+        ignored_params: Set[nn.Parameter],
+    ) -> None:
+        """
+        Raises an error if ``module`` has original parameters on multiple
+        devices, ignoring the parameters in ``ignored_params``. Thus, after
+        this method, the module must be either fully on the CPU or fully on a
+        non-CPU device.
+        """
+        devices = set(
+            param.device for param in self._get_orig_params(module, ignored_params)
         )
+        if len(devices) > 1:
+            raise RuntimeError(
+                f"FSDP only supports single device modules but got params on {devices}"
+            )
+
+    def _get_compute_device(
+        self,
+        module: nn.Module,
+        ignored_params: Set[nn.Parameter],
+        device_id: Optional[Union[int, torch.device]],
+    ) -> torch.device:
+        """
+        Determines and returns this FSDP instance's compute device. If the
+        module is already on a non-CPU device, then the compute device is that
+        non-CPU device. If the module is on CPU, then the compute device is the
+        current device.
+
+        The ``device_id`` argument should be what the user passed into the FSDP
+        constructor to produce a faithful error. If ``device_id`` is not
+        ``None``, then we enforce that it must either be (1) the same as the
+        compute device, (2) the same index as the compute device's index, or
+        (3) the general unindexed device. Hence, ``device_id`` is simply a way
+        for the user to have FSDP move the module to device, but that target
+        device is fixed to the current device.
+
+        NOTE: For now, the compute device is always a CUDA GPU device with its
+        explicit index.
+
+        Precondition: ``_check_single_device_module()``.
+        """
+        if not torch.cuda.is_available():
+            raise RuntimeError("FSDP does not support CPU only execution")
+        # If the module is on GPU already, then that GPU device has priority
+        # over the current device
+        param = next(self._get_orig_params(module, ignored_params), None)
+        if param is not None and param.device.type == "cuda":
+            compute_device = param.device
+        else:
+            compute_device = torch.device("cuda", torch.cuda.current_device())
```

Comment:
not completely sure if this is BC, because now we may throw errors if user did not call torch.cuda.set_device but specifies device_id, but we can probably just start enforcing user calls torch.cuda.set_device before FSDP ctor, it is best practice

Reply:
The meta device default initialization might be problematic if the user does not call `torch.cuda.set_device()`:
https://github.com/pytorch/pytorch/blob/1e4383f7563c9bb2e3c8e6989b6853d1d04f652f/torch/distributed/fsdp/fully_sharded_data_parallel.py#L119